solr/core/src/java/org/apache/solr/cluster/placement/plugins/SamplePluginMinimizeCores.java

```diff
@@ -0,0 +1,132 @@
+ * limitations under the License.
+ */
+
+package org.apache.solr.cluster.placement.plugins;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.Map;
+
+import com.google.common.collect.Ordering;
+import com.google.common.collect.TreeMultimap;
+import org.apache.solr.cluster.placement.Cluster;
+import org.apache.solr.cluster.placement.CoresCountPropertyValue;
+import org.apache.solr.cluster.placement.CreateNewCollectionPlacementRequest;
+import org.apache.solr.cluster.placement.Node;
+import org.apache.solr.cluster.placement.PlacementException;
+import org.apache.solr.cluster.placement.PlacementPlugin;
+import org.apache.solr.cluster.placement.PropertyKey;
+import org.apache.solr.cluster.placement.PropertyKeyFactory;
+import org.apache.solr.cluster.placement.PropertyValue;
+import org.apache.solr.cluster.placement.PropertyValueFetcher;
+import org.apache.solr.cluster.placement.Replica;
+import org.apache.solr.cluster.placement.ReplicaPlacement;
+import org.apache.solr.cluster.placement.PlacementRequest;
+import org.apache.solr.cluster.placement.PlacementPlan;
+import org.apache.solr.cluster.placement.PlacementPlanFactory;
+import org.apache.solr.common.util.SuppressForbidden;
+
+/**
+ * Implements placing replicas to minimize number of cores per {@link Node}, while not placing two replicas of the same
+ * shard on the same node.
+ *
+ * TODO: code not tested and never run, there are no implementation yet for used interfaces
+ */
+public class SamplePluginMinimizeCores implements PlacementPlugin {
+
+  @SuppressForbidden(reason = "Ordering.arbitrary() has no equivalent in Comparator class. Rather reuse than copy.")
+  public PlacementPlan computePlacement(Cluster cluster, PlacementRequest placementRequest, PropertyKeyFactory propertyFactory,
+                                        PropertyValueFetcher propertyFetcher, PlacementPlanFactory placementPlanFactory) throws PlacementException {
+    // This plugin only supports Creating a collection.
+    if (!(placementRequest instanceof CreateNewCollectionPlacementRequest)) {
+      throw new PlacementException("This toy plugin only supports creating collections");
+    }
+
+    final CreateNewCollectionPlacementRequest reqCreateCollection = (CreateNewCollectionPlacementRequest) placementRequest;
+
+    final int totalReplicasPerShard = reqCreateCollection.getNrtReplicationFactor() +
+        reqCreateCollection.getTlogReplicationFactor() + reqCreateCollection.getPullReplicationFactor();
+
+    if (cluster.getLiveNodes().size() < totalReplicasPerShard) {
+      throw new PlacementException("Cluster size too small for number of replicas per shard");
+    }
+
+    // Get number of cores on each Node
+    TreeMultimap<Integer, Node> nodesByCores = TreeMultimap.create(Comparator.naturalOrder(), Ordering.arbitrary());
```

Comment:
I believe the property fetching is overly complicated . We should probably make it a lot simpler. 

Basically, the only requirement is strong typing. 

`TreeMultimap<Integer, Node> nodesByCores = TreeMultimap.create(Comparator.naturalOrder(), Ordering.arbitrary());`

This definitely is not the easiest code we could write. A user just wants to get an integer value for the # of cores in a node.

Reply:
Having 2 classes for each property will just multiply the no:of classes in the API. Basically

The only difference between 2 properties is property name and the type. Property name can be supplied as a method, type can be a generic

Reply:
Strong typing was one requirement (and strong typing is not only return type, it's also nice names for methods to access these values), the other one was efficiency: being able to fetch ALL needed properties from a remote Node at once (i.e. not requesting metrics first and system properties second which would force the Solr side framework to do two round trips - caching and optimizations aside).
That's why we have the notion of property key, to be able to "accumulate" the properties to fetch, then we have asingle interface per property, allowing clean access to the value (for cores it's the call coresCountPropertyValue.getCoresCount() line 85 below).

The "complexity" of the structure here is the plug-in implementation, not the API, using sorted data structures to compute a placement quickly. This should be compared (in complexity) to the previous approach of generating all possible placements and comparing them.

Reply:
I agree with @murblanc about the need for efficient multi-fetching. But I also agree with @noblepaul :) that we could partially relax strong-typing requirements here in order to drastically reduce the API surface.

Reply:
I challenge the "drastically reduce" assertion here @sigram. I think we can save a couple or so lines of code per property type, but maybe not much more. We can likely simplify the `PropertyValue` hierarchy to a few typed abstractions, see section at the end of [this comment](https://github.com/apache/lucene-solr/pull/1684#issuecomment-684063923).
If we do keep the efficiency of multi fetching (all properties from a node at once) and want to hide from the plugin implementer the ability to fetch from multiple nodes concurrently, we do need to keep this whole logic of keys totally describing what they refer to.